transloadit: fix progress with very different Assembly runtimes

agreene-coursera opened the pull request:
…se it was smaller than other files in the batch), complete it during post processing

Durning an internal QA of our migration to the most recent version of Uppy, we came across a bug during large batch uploads using the transloadit plugin with several assemblies. In some cases if the file sizes in the different assemblies were dramatically different, the smaller uploads would have their assembly complete before the rest of the files finished uploading. This would mean that during the postProcessing step, the assemblyWatcher would never receive a `assembly-complete`event for that assembly so the step would hang.

This is a naive attempt to remedy the bug by keeping track of what assemblies have completed and when we start postProcessing, emiting `postprocess-complete `for any that have already finished up.

agreene-coursera commented:
Hmm not sure why the travis build is still failing. I removed the offending semi-colon that was causing the linter to complain and I'm pretty sure this shouldn't effect the companions

goto-bus-stop commented:
Thanks for the PR! This approach seems reasonable. The Companion tests are flaky at the moment so those failures can be ignored.

I'll look at this more closely on Monday, I want to make sure that there isn't another similar case happening for another type of event. If there is we may need to create the AssemblyWatcher instances in the `prepareUpload` phase already.

goto-bus-stop commented:
Hmm even with this patch I still sometimes get files hung in the postprocess step unfortunately. For example with this code that creates one short-running assembly and one with many files:
```js
let isFirst = true
uppy.use(Transloadit, {
  waitForEncoding: true,
  getAssemblyOptions (file) {
    // force creation of multiple assemblies
    let dummyField = 'not first'
    if (isFirst) {
      isFirst = false
      dummyField = 'first'
    }
    return {
      params: {
        auth: { key: TRANSLOADIT_KEY },
        template_id: TRANSLOADIT_TEMPLATE,
        fields: { dummyField }
      }
    }
  }
})
```
And then uploading like 60 photos with a trivial template (just a fillcrop or something), the first file may still hang. I'm not sure why—I expected the `finishedAssemblyIDs` approach would do the trick here!

I do think we may need to create the AssemblyWatcher earlier or maybe rely on fetching the assembly status from the API in some way. I can do further debugging this week too.

agreene-coursera commented:
@goto-bus-stop that's super strange, I can't think of a reason why that would still happen... I can give moving the AssemblyWatcher instantiation around a shot

agreene-coursera commented:
@goto-bus-stop looks like that strategy works for all of our test cases that were broken before, let me know what you think

goto-bus-stop commented on packages/@uppy/transloadit/src/index.js:
```diff
@@ -251,6 +254,42 @@ module.exports = class Transloadit extends Plugin {
+  // AssemblyWatcher tracks completion states of all Assemblies in this upload.
+    const watcher = new AssemblyWatcher(this.uppy, assemblyID)
+
+    watcher.on('assembly-upload', (id) => {
```
When I try it with this PR and `waitForEncoding: true`, the postprocessing step is never entered. I think it's because the `assembly-upload` event is not actually fired anywhere right now.

goto-bus-stop replied:
I think the most _accurate_ moment to fire that event would be when all the files belonging to an assembly have finished uploading. An easier way would be to fire it when `uppy.on('transloadit:assembly-executing')` fires, which is when Transloadit tells us that it has received all the files. That's typically going to be a few ms later but it might be OK.

agreene-coursera commented:
Whoops my last commit didn't make it into the PR. I moved away from firing postproccessing-progress in the watcher since it caused a bit of UI jank. Instead I switched to calling it in afterUpload but only for files that haven't already completed

goto-bus-stop commented:
Ahh nice that looks good! I'll push a few minor tweaks:
- Keep track of `completedFiles` by their ID string, since file objects are immutable and can be cloned frequently, so their reference identity cannot be relied on
- Add back the `{ mode, message }` object parameter to `postprocess-progress` so the "encoding..." progress bar shows up when `waitForEncoding: true` is in use

I'll run a few more manual tests in setups with Golden Retriever and other plugins to make sure the interactions there are all fine. Hopefully we can merge+release later today :)

arturi commented:
@agreene-coursera This is out in the latest @uppy/transloadit and `@uppy/robodog`! Thanks!

agreene-coursera commented:
Awesome! Thanks @goto-bus-stop and @arturi!
